### ISSUE TITLE
Type inference fails when type narrowing occurs within a loop

### ISSUE THREAD

**Issue by kishore03109** [microsoft/TypeScript#62279]
### 🔎 Search Terms

Type inference following a type narrowing fails when done within a loop. 

### 🕗 Version & Regression Information

This bug exists ever since TS 3.7 where TS asserts was introduced. 

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4A8AKgHwAUAUHHGpddigFxx6AGlYBKAYWKwCnbjVwBvdnEwI4zAIRcUVeaLhKOHGAAsoEAO5wUwKwFEo5qM1HKAvqw+sYATzDA4ADc8ABtMCno-YBkAXjgAcgBBeLgAHwSAIRTWVm0CeHy8GCQCAWCwiKjYhOSc5HR5OF9-AEkUBBJgdGAAMTxMMJQAczaCcOAAGQgIMBcDZUkSGHIdHhxmQuKZDRi4+KzXDgRodTz4AH04CDUAbQBGIQAmAF19QyMAeg+4AFUbcGAGGAFDgJCcAnimxKjxSmAAtmAwmhqCEfHATIQmlEEngUD4UgAjQF4EoBahwCgQaLWCDwDGBAJ4LH+fAoFC0oq8VkgzAyKDADr87ogqj8jCoq5QFSrMUwCWYXDUGSWRUoHihTAALxIADplBwznAoQRHnA4sa4Pr8EQlitdLwNjAitCzbsavFDnAPF56hguc1gG1BV00MAAAqSBVDADq1BMECQMCmMzm70WZDkDotOz2B2UhuNpvNTq2lo46eWmfWhdde2SrjcQA

### 💻 Code

```ts
export function assertCondition<T>(
  condition: T,
): asserts condition {
  if (!condition) {
    throw new Error()
  }
}

type validTypes = 'A' | 'B' 

const status: validTypes = 'A'

function typeInferenceFailingInsideLoop() {
  assertCondition(status !== 'B')
  for (const _ of [1,2]) {
    // Unexpected error: 'status2' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
    const status2 = status 
    assertCondition(status2 === 'A')
  }
}

function typeInferencePassingWithoutLoop() {
  assertCondition(status !== 'B')
  const status2 = status 
  assertCondition(status2 === 'A')  
}
```

### 🙁 Actual behavior

Error thrown within the function call `typeInferenceFailingInsideLoop` stating `'status2' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

### 🙂 Expected behavior

Ideally, no type error should be thrown here. 

#### Possible workarounds 
1. Redeclare the type to help TS infer ([Playground link](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4A8AKgHwAUAUHHGpddigFxx6AGlYBKAYWKwCnbjVwBvdnEwI4zAIRcUVeaLhKOHGAAsoEAO5wUwKwFEo5qM1HKAvqw+sYATzDA4ADc8ABtMCno-YBkAXjgAcgBBeLgAHwSAIRTWVm0CeHy8GCQCAWCwiKjYhOSc5HR5OF9-AEkUBBJgdGAAMTxMMJQAczaCcOAAGQgIMBcDZUkSGHIdHhxmQuKZDRi4+KzXDgRodTz4AH04CDUAbQBGIQAmAF19QyMzuE2Sx7LQ8Mi-mq3xkyg4izIcl4GxgRR+cF2e2ShzgHi89QwvCaUTaHSgXTQwAACpJMMMAOrUEwQJAwKYzObvCHLKHrEFwHZ7A7KT4gx4Ir6wrZwBZEJYrXTQvkI3Y1eKuNxAA)) Interestingly, the type narrowing still works as intended. 

2. Don't use generic ([Playground link](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4AUal12KAXHAEYQQA2weKASjaFisAnAYoqNXAG8AUHDiYEcWgEJJ05gLgKlSmAAsoEAO5wUwCwFEopqLQEBuRXAC+8z-JgBPMMBwAG543JgUACr+wOIAvHAA5ACCCXAAPokAQqkZqBTACJhWFHDy8pIE8JV4MEgEbCFhkdFxiSllyOgycH4BAJIoCCTA6MAAYniYYSgA5gME4cAAMlxgTnpuIiQw5FJMdOrq1bUEAm4I0GoV8AD6cBCqANoAjAA0AEwAurr6Btdwxzq7zg8UB4jcSi2ZEYMloYOBsURbQSZyUnm8nQwzB60QGQygIzQwAACiIijMAOrUIwQJAwFYQNY-TZEba7bQHI4wGp1VESHCVAHck4IoU88GQ1nQvaw+EgpHJFFeIA)) For checking of truthy conditions, one would need to change this to a boolean condition instead.

**Comment by Andarist** [comment]
Your `assertCondition` doesn't really have to generic, just use this ([TS playground](https://www.typescriptlang.org/play/?ts=6.0.0-dev.20250814#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4AUal12KAXHAEYQQA2weKASjaFisAnAYoqNXAG8AUHDiYEcWgEJJ05gLgKlSmAAsoEAO5wUwCwFEopqLQEBuRXAC+8z-JgBPMMBwAGp43JgUACr+gQC8cABEAILxcAA+CQBCKenxpPGu8hTAaNx4UIGSBPBVeDBIBGwhYZHRBcjoMnAwwFUAjE56biIkMORSTHQ1deLqMXHxWS5uldUwtfUATHBxU-WuSsNkjDK0uwRbc-PJS0pnG-twAPSPcAB6APxe8vLtGMxdPRgGwG+nwRBGY20kzW0zgs3miweCGgahWcAA+nAIKoANq9AA0cA2AF1dKClGi7ts4GcHgdwUdxicqZcEtc6TSYZsOc83p8lJ5PEA)):
```ts
export function assertCondition(condition: boolean): asserts condition {
  if (!condition) {
    throw new Error();
  }
}
```

**Comment by kishore03109** [comment]
Hey @Andarist, unfortunately `assertCondition` is a generic because I wanted to take in truthy conditions

eg. 
```
type validTypes = 'A' | 'B' | undefined

const status: validTypes = 'A'

function typeInferenceFailingInsideLoop() {
  assertCondition(status) // don't wish to change this to assertCondition(!!status)
  for (const _ of [1,2]) {
    // Unexpected error: 'status2' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. 
    const status2 = status 
    assertCondition(status2 === 'A')
  }
}
```


Have added this as a possible workaround in the PR description for others to refer to.

**Comment by Andarist** [comment]
> unfortunately assertCondition is a generic because I wanted to take in truthy conditions

I don't understand what you mean here. The version proposed by me only accepts `boolean`s as arguments. You won't get `true` as an logical expression's type even if it's known to be `true` at runtime:
```ts
const result = true === true // boolean
```

**Comment by kishore03109** [comment]
[Here is a playground link](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV2pgc0xQEMAbAQQGcrhYBhHAE0yxwB4AVAPgAoAoOHDTNW2FAC44nADT8AlFJI06MKsNFtcAb0FxMCOLwCEIlCy3y4uoUJgALKBADucFMFcBRKE6i95egC+-MH8oJCwiKgY4m4e1LQMmuK8ZhbiUgBGEBBkwCQoinDKiWoa5mI41noGRqbJOFY2tg5Oru5ePtD+ANxBIfz8MACeYMBwAG7kmEyco8DqALxwAOQUK3AAPqsAQhvbqEzACMTATINmVPBXJDBIVFJTZDNzY0ur64PI6FrFKkkVLSeEAkAC2YDy-mqQnwRFIlH+MEYgJSNzuVCsAHpMXAAGrTJhwPQdBKqZHpHC8NH3LE417jby+GRxM5wGAQOD3cYkxHkyooEzGakY-QoK75c56BDQIyXeAAfTwhgA2gBGGQAJgAuk09EI5XBhRq4MthUTbHFnKSARSBUaTYtlmsVrS4M5oABrEhOQ5uz1UfrBIA) to further demonstrate my point. Your workaround definitely works, but it does limit the types of arguments that I can call `assertCondition` with. A bit of context is that this example is extracted from a code base where the function `assertCondition` is used quite exhaustively. Thus, I prefer [this workaround](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4A8AKgHwAUAUHHGpddigFxx6AGlYBKAYWKwCnbjVwBvdnEwI4zAIRcUVeaLhKOHGAAsoEAO5wUwKwFEo5qM1HKAvqw+sYATzDA4ADc8ABtMCno-YBkAXjgAcgBBeLgAHwSAIRTWVm0CeHy8GCQCAWCwiKjYhOSc5HR5OF9-AEkUBBJgdGAAMTxMMJQAczaCcOAAGQgIMBcDZUkSGHIdHhxmQuKZDRi4+KzXDgRodTz4AH04CDUAbQBGIQAmAF19QyMzuE2Sx7LQ8Mi-mq3xkyg4izIcl4GxgRR+cF2e2ShzgHi89QwvCaUTaHSgXTQwAACpJMMMAOrUEwQJAwKYzObvCHLKHrEFwHZ7A7KT4gx4Ir6wrZwBZEJYrXTQvkI3Y1eKuNxAA) of just re-declaring the type again.

**Comment by Andarist** [comment]
Oh, in hindsight it seems obvious what you have meant 😅

**Comment by RyanCavanaugh** [comment]
Without a more concrete reason to write `assertCondition` as a generic function, I don't really think this is a bug worth slowing down other mainline codepaths for (`checkNodeDeferred` incurs a closure allocation and should really only be used sparingly). You could equivalently write
```ts
export function assertCondition(condition: unknown): asserts condition {
```
to take an arbitrary value. I can't think of any reason to put a type parameter there that doesn't *also* involve making `assertCondition` an invalid assertion function.

**Comment by Andarist** [comment]
Note that I managed to repro this problem without using assertion functions, this can happen with regular inferrrable calls: [TS playground](https://www.typescriptlang.org/play/?ts=6.0.0-dev.20250821#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXy1URBgGUNlFEAeAFQD4AKWAcwC55aAaeMAC3ABrcpUQB+DgCMcOCCCioAlB1oBuAFDqQADwAOOGBiRpMufBhABnDACZGEQlY7WYhFgG0Ausvgu3X+ABvdXh4OSMUdA5g0NDUKABbEGcMV1QWDVCAX3gAH3hUZAgIeABeAqKIDRCkA3hGMDxreAB9eBxEMMdLRSCa0KxOxgBZKAw+ADoYBWAcBMZe6ngABgmAVl6Y2NDIsDK+7e34pI4AIgBHAHcSU65+2KzMh-gQCEsELe3B+t3N++3GqhmoECokEDlyrsnodjOh9oRiGQKFRGP9Dp8YbFjsl4KccOMbndMQ8icTQUkyqVyqcoJIwLc0aFFNCHvcsjV2TU4BQYPgHKgrBp2UA). I have that test case in https://github.com/microsoft/TypeScript/pull/60865

> `checkNodeDeferred` incurs a closure allocation and should really only be used sparingly

I don't see any closure allocations associated with `checkNodeDeferred` 🤔

**Comment by kishore03109** [comment]
@RyanCavanaugh what about the use case where `assertCondition` cannot take in any Promises during build time? 
eg.

```
export function assertCondition<T>(
  condition: T extends Promise<unknown> ? never : T,
): asserts condition {
  if (!condition) {
    throw new Error()
  }
}
```

[TS playground link](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwIYDOBwsAwjgCaZY4A8AKgHwAUAUHHGpddigFxx6cUDGAoKBOAAUoEALaZitVAGsUEAO4pGcAPxwUwAG4k4A+gBpWASgGFisSV3E8ccAN7s4mBHGYBCZyoaFGsPLw4YAAtZDQNgOIBRKFkoZmsvAF9WbNYYAE8wYDgjPAAbTAp6QuBJAF44AHIAQUa4AB8mgCE21lZnAnhBvBgkAgFSiqqa+qbWvsJ89ERUDF44AqKASRQEEjE0YAAxPEwKlABzHYJK4AAZCAgwdPCOexIYchcQ5mHRyX8dQajR6GQ4CGgfgG8AA+nAIL4ANoARgsACYALphTwcDjQuB-MZoiblSrVIqzQmSCL4IgfL7BXi-GAjIlwIHA1pguDZPE4QZwEDs+JxGTyRTAZhQWrsnTSgjc95kbg-PAaU7wECKunK75MrVwAD0hsEUUUBKiECQZQoGximnwuBIqRyfWQ6BCGxqOz20vQwCk9kwlwA6tRLUgYA8ni8cbSHJ8VUyqXBAcDQV58VS0cKU14lYm9ThmayCDmOXNGhlMkA)

**Comment by RyanCavanaugh** [comment]
TypeScript doesn't support negated types, so workarounds to try to make negated types are going to necessarily encounter some limitations
